Zephyr sdk improvement

tejlmand opened the pull request:
This PR is a Zephyr repo followup to this commit https://github.com/zephyrproject-rtos/sdk-ng/pull/199 and is dependent on that PR to be merged, and a new Zephyr SDK Toolchain with CMake package to be released.

This commit introduces Zephyr SDK CMake config package.

This removes the need for setting ZEPHYR_SDK_INSTALL_DIR and
ZEPHYR_TOOLCHAIN_VARIANT when using Zephyr SDK in Linux.
It also allows to introduces never SDKs without breaking Zephyr.

For example, with this PR, the current Zephyr SDK is 0.11.1 but when
releasing 0.12 then the current Zephyr will no longer built.
This PR moves the Zephyr SDK CMake related code to the SDK and thus
allowing to use newer SDKs, as long as they are backwards compatible.

It also allows multiple SDK installations, and will automatically select
the version closet to the required version.

---
Additional info:

Today, when doing a bisect or jumping to an LTS release, for example Zephyr 1.14.1, but having a Zephyr-SDK `0.11.1` will give an error similar to this:
```
tora@ubuntu /p/g/n/z/s/hello_world> rm -rf build; cmake -GNinja -DBOARD=nrf51_pca10028 -Bbuild .
Zephyr version: 2.1.0
-- Found PythonInterp: /usr/bin/python3 (found suitable version "3.6.8", minimum required is "3.4") 
-- Selected BOARD nrf51_pca10028
CMake Error at /projects/github/ncs/zephyr/cmake/toolchain/zephyr/host-tools.cmake:57 (include):
  include could not find load file:

    /projects/github/ncs/zephyr/cmake/toolchain/zephyr/0.11/host-tools.cmake
Call Stack (most recent call first):
  /projects/github/ncs/zephyr/cmake/host-tools.cmake:3 (include)
  /projects/github/ncs/zephyr/cmake/app/boilerplate.cmake:443 (include)
  CMakeLists.txt:5 (include)


-- Found west: /home/tora/.local/bin/west (found suitable version "0.7.2", minimum required is "0.6.0")
CMake Error at /projects/github/ncs/zephyr/cmake/toolchain/zephyr/generic.cmake:7 (include):
  include could not find load file:

    /projects/github/ncs/zephyr/cmake/toolchain/zephyr/0.11/generic.cmake
Call Stack (most recent call first):
  /projects/github/ncs/zephyr/cmake/generic_toolchain.cmake:64 (include)
  /projects/github/ncs/zephyr/cmake/app/boilerplate.cmake:459 (include)
  CMakeLists.txt:5 (include)


-- Loading /projects/github/ncs/zephyr/boards/arm/nrf51_pca10028/nrf51_pca10028.dts as base
CMake Error at /projects/github/ncs/zephyr/cmake/dts.cmake:130 (message):
  command failed with return code: No such file or directory
Call Stack (most recent call first):
  /projects/github/ncs/zephyr/cmake/app/boilerplate.cmake:460 (include)
  CMakeLists.txt:5 (include)
```
or
```
tora@ubuntu /p/g/n/z/s/hello_world> rm -rf build; cmake -GNinja -DBOARD=nrf51_pca10028 -Bbuild .
Zephyr version: 1.14.1
-- Found PythonInterp: /usr/bin/python3 (found suitable version "3.6.8", minimum required is "3.4") 
-- Selected BOARD nrf51_pca10028
CMake Error at /projects/github/ncs/zephyr/cmake/toolchain/zephyr/host-tools.cmake:47 (include):
  include could not find load file:

    /projects/github/ncs/zephyr/cmake/toolchain/zephyr/0.11/host-tools.cmake
Call Stack (most recent call first):
  /projects/github/ncs/zephyr/cmake/host-tools.cmake:3 (include)
  /projects/github/ncs/zephyr/cmake/app/boilerplate.cmake:382 (include)
  CMakeLists.txt:5 (include)


-- Found west: /home/tora/.local/bin/west (found suitable version "0.7.2", minimum required is "0.5.6")
CMake Error at /projects/github/ncs/zephyr/cmake/toolchain/zephyr/generic.cmake:7 (include):
  include could not find load file:

    /projects/github/ncs/zephyr/cmake/toolchain/zephyr/0.11/generic.cmake
Call Stack (most recent call first):
  /projects/github/ncs/zephyr/cmake/generic_toolchain.cmake:57 (include)
  /projects/github/ncs/zephyr/cmake/app/boilerplate.cmake:398 (include)
  CMakeLists.txt:5 (include)


-- Loading /projects/github/ncs/zephyr/boards/arm/nrf51_pca10028/nrf51_pca10028.dts as base
-- Overlaying /projects/github/ncs/zephyr/dts/common/common.dts
CMake Error at /projects/github/ncs/zephyr/cmake/dts.cmake:126 (message):
  command failed with return code: No such file or directory
Call Stack (most recent call first):
  /projects/github/ncs/zephyr/cmake/app/boilerplate.cmake:399 (include)
  CMakeLists.txt:5 (include)


-- Configuring incomplete, errors occurred!

```


This PR removes the need for setting `ZEPHYR_SDK_INSTALL_DIR` or `ZEPHYR_TOOLCHAIN_VARIANT` if installing Zephyr SDK under one the following locations:
- `${HOME}/zephyr-sdk*`
- `/opt/zephyr-sdk*`
- `/usr/zephyr-sdk*`
- `/usr/local/zephyr-sdk*`
where the `*` denotes that anything can be added after the beginning, that is:
`/opt/zephyr-0.11.1` and `/opt/zephyr-sdk` are both valid.
This means that in Zephyr repo the following code is added, and Zephyr uses `find_package(Zephyr-sdk VERSION)` to lookup Zephyr.

This has the following advantages:
1. Installing multiple Zephyr sdk's without changing environment
1. Source the additional cmake files required from the Zephyr SDK instead, making it possible to use a future Zephyr SDK with older build (the SDK decides if it is backward compatible)
1. Zephyr build will select the best matching version, to the minimum required.

As example, having a newer Zephyr SDK `0.12.0` but no `${MAJOR_MINOR}` version in includes will now work, for example:
```
 find_package(Zephyr-sdk 0.11.1)
```
will result in `0.12.0` being used, if that is only Zephyr SDK installed (and is compatible):
```
-- Found toolchain: zephyr (/opt/zephyr-sdk-0.12.0)
```
but let's say user only has `0.13.0` and this one is not compatible with `0.11.1` requirement, then the following message is printed:
```
  The SDK version you are using is not supported, please update your SDK.

  You need at least SDK version 0.11.1.

  You have version(s):

    0.10.0 (/opt/zephyr-sdk-0.10.0)
    0.13.0 (/opt/zephyr-sdk-0.13.0)

  The new version of the SDK can be downloaded from:
  
  https://github.com/zephyrproject-rtos/sdk-ng/releases/download/v0.11.1/zephyr-sdk-0.11.1-setup.run
```
which is much better user experience than today's cryptic message.

tejlmand committed:
cmake: Zephyr sdk package handling

This commit introduces Zephyr SDK CMake config package.

This removes the need for setting ZEPHYR_SDK_INSTALL_DIR and
ZEPHYR_TOOLCHAIN_VARIANT when using Zephyr SDK in Linux.
It also allows to introduces never SDKs without breaking Zephyr.

For example, with this PR, the current Zephyr SDK is 0.11.1 but when
releasing 0.12 then the current Zephyr will no longer built.
This PR moves the Zephyr SDK CMake related code to the SDK and thus
allowing to use newer SDKs, as long as they are backwards compatible.

It also allows multiple SDK installations, and will automatically select
the version closet to the required version.

Signed-off-by: Torsten Rasmussen <Torsten.Rasmussen@nordicsemi.no>

tejlmand commented:
@mnkp This is now done: https://github.com/zephyrproject-rtos/zephyr/pull/23054#discussion_r399518948
> > I actually think we might be able to use the CMake package system for the toolchain as well, so that only user's who has special needs has to set `ZEPHYR_TOOLCHAIN_VARIANT` and `ZEPHYR_SDK_INSTALL_DIR` themselves, for example when using a 3rd party toolchain.
> 
> That would be nice!

No need to set those environment variables, if this PR goes in together with this: https://github.com/zephyrproject-rtos/sdk-ng/pull/199

tejlmand commented:
@nashif @galak @carlescufi  Shippable fails, cause the current Zephyr toolchain is not a CMake package, so this is expected at current stage.

mbolivar-nordic commented:
I fully support the goals of this pull request.

A couple of small remarks from reading the description:

> ${HOME}/zephyr-sdk*
> /opt/zephyr-sdk*
> /usr/zephyr-sdk*
> /usr/local/zephyr-sdk*

On Linux, I think it would be very nice to allow `~/.local/opt/zephyr-sdk*` at least. Some users (myself included) prefer to keep things outside of their home directory inside ~/.local where possible, as a tidiness thing.

Is that possible?

>> but let's say user only has 0.13.0 and this one is not compatible with 0.11.1 requirement, then the following message is printed:
>   You need at least SDK version 0.11.1.

I find this hard to understand, because 0.13.0 > 0.11.1 as a comparison between semantic versions, so as a user I would say "I **do** have an SDK at least that version!"

Is it possible to make the error message say "You need an SDK compatible with 0.11.1" instead?

tejlmand commented:
> A couple of small remarks from reading the description:
> 
> > ${HOME}/zephyr-sdk*
> > /opt/zephyr-sdk*
> > /usr/zephyr-sdk*
> > /usr/local/zephyr-sdk*
> 
> On Linux, I think it would be very nice to allow `~/.local/opt/zephyr-sdk*` at least. Some users (myself included) prefer to keep things outside of their home directory inside ~/.local where possible, as a tidiness thing.
> 
> Is that possible?
> 

yes, will add those two:
- `~/.local/opt/zephyr-sdk*`
- `~/.local/zephyr-sdk*`

Do you think we should have more default places ?

> > > but let's say user only has 0.13.0 and this one is not compatible with 0.11.1 requirement, then the following message is printed:
> > 
> > 
> > You need at least SDK version 0.11.1.
> 
> I find this hard to understand, because 0.13.0 > 0.11.1 as a comparison between semantic versions, so as a user I would say "I **do** have an SDK at least that version!"
> 

yeah, I agree, will fix.
This was actually taken from old help text, where it only printed if you had to old version, cause to new version simply just break CMake.

> Is it possible to make the error message say "You need an SDK compatible with 0.11.1" instead?

of course it is.

tejlmand commented:
> > On Linux, I think it would be very nice to allow `~/.local/opt/zephyr-sdk*` at least. Some users (myself included) prefer to keep things outside of their home directory inside ~/.local where possible, as a tidiness thing.
> > Is that possible?
> 
> yes, will add those two:
> 
>     * `~/.local/opt/zephyr-sdk*`
> 
>     * `~/.local/zephyr-sdk*`
> 

Added.
> > > > but let's say user only has 0.13.0 and this one is not compatible with 0.11.1 requirement, then the following message is printed:
> > > 
> > > 
> > > You need at least SDK version 0.11.1.
> > 
> > 
> > I find this hard to understand, because 0.13.0 > 0.11.1 as a comparison between semantic versions, so as a user I would say "I **do** have an SDK at least that version!"
> > 
> > Is it possible to make the error message say "You need an SDK compatible with 0.11.1" instead?
> 
> of course it is.

Now printing:
```
  You need SDK version 0.11.1 or newer.

  You have version(s):

    0.10.0 (/opt/zephyr-sdk-0.10.0)
```
when only having to old version (which will be most common case).

or:
```
  You need SDK version 0.11.1 or compatible version.

  You have version(s):

    0.10.0 (/opt/zephyr-sdk-0.10.0)
    0.13.0 (/opt/zephyr-sdk-0.13.0)
```
when you have a newer version installed, than required, and that version was not compatible.

mbolivar-nordic commented:
> Do you think we should have more default places ?
>

That's perfect for me. Thank you.



> when you have a newer version installed, than required, and that version was not compatible.

This looks great to me. Thanks for the changes!

galak commented:
DEV-REVIEW (april 16):

@nashif - need to maintain existing functionality and documenting
@galak - look at adding .github workflow that verifies this works using both the cmake package method and env method [as a followup]

tejlmand commented:
> DEV-REVIEW (april 16):
> 
> @nashif - need to maintain existing functionality and documenting

Thanks for the concern, it's always good to double check that everything is in place.
This PR contains those sections in:
https://github.com/tejlmand/zephyr/blob/zephyr_sdk_improvement/doc/getting_started/installation_linux.rst

> It is recommended to install the Zephyr SDK at one of the following locations:
>
>  ........
> 
> If you install the Zephyr SDK outside any of those locations, then it is required to set :envvar:`ZEPHYR_SDK_INSTALL_DIR` to point to the Zephyr SDK installation folder.

and this:
> As already noted above, the SDK also includes prebuilt host tools.  To use the 
> SDK's prebuilt host tools with a toolchain from another source, you must set the
> :envvar:`ZEPHYR_SDK_INSTALL_DIR` environment variable to the Zephyr SDK
> installation directory. To build without the Zephyr SDK's prebuilt host tools, the :envvar:`ZEPHYR_SDK_INSTALL_DIR` environment variable must be unset.

@nashif Let me know if this does not fulfil your expectations.

carlescufi commented on cmake/toolchain/zephyr/host-tools.cmake:
```diff
@@ -7,51 +7,77 @@
+                 /usr
+		 /usr/local
+		 /opt
+		 $ENV{HOME}
```
Add `$ENV{HOME}/bin`

tejlmand replied:
added

carlescufi commented on cmake/toolchain/zephyr/host-tools.cmake:
```diff
@@ -7,51 +7,77 @@
+  # to the required, as example, requesting 0.11.1:
+  # Zephyr-sdk: 0.10.0  (Not matching)
+  # Zephyr-sdk: 0.11.1  (First matching - will be used)
+  # Zephyr-sdk: 0.12.0  (Matching)
```
@galak and @nashif if the required version is 0.11.1, and the user has the SDK versions installed above an has not set `ZEPHYR_SDK_INSTALL_DIR`, should we use 0.11.1 or 0.12.0?

tejlmand replied:
I have no strong preference, but the reason I chose this order is that whenever we release Zephyr, then we are testing with the minimum required version, as that will usually be the only working SDK available at the time of release.
This again means this particular SDK version is assumed to be fully working with this Zephyr release.

When we release a new SDK (for example 0.11 -> 0.12), then that version is of course expected to be backward compatible, but as we have not fully tested all corners of a new SDK with the old release, then I decided the safest bet was to select the closest version when multiple versions are available.

But choosing newest compatible version is also good choice.
I just had to present one way or the other in this PR, and the reason above states why this is the initial proposal.

carlescufi replied:
Dev-review: Use the latest version instead of the required one.

tejlmand replied:
fixed

carlescufi commented on cmake/toolchain/zephyr/host-tools.cmake:
```diff
@@ -7,51 +7,77 @@
 
-set_ifndef(ZEPHYR_SDK_INSTALL_DIR "$ENV{ZEPHYR_SDK_INSTALL_DIR}")
-set(ZEPHYR_SDK_INSTALL_DIR ${ZEPHYR_SDK_INSTALL_DIR} CACHE PATH "Zephyr SDK install directory")
+set(REQUIRED_SDK_VER 0.11.1)
```
Should this be renamed to `MIN_REQUIRED_SDK_VER`?

tejlmand replied:
Updated to: `MINIMUM_REQUIRED_SDK_VERSION`

carlescufi commented on doc/getting_started/index.rst:
```diff
@@ -346,23 +346,32 @@ including: compiler, assembler, linker, and their dependencies.
+            * ``$HOME/.local/opt/zephyr-sdk[-x.y.z]``
+            * ``/opt/zephyr-sdk[-x.y.z]``
+            * ``/usr/zephyr-sdk[-x.y.z]``
+            * ``/usr/local/zephyr-sdk[-x.y.z]``
```
Add `$HOME/bin/zephyr-sdk[-x.y.z]`

tejlmand replied:
We can add all the search folders we like, but question is if we should add this folder to recommended list.

Personally, I think the `bin` folder is for the executables (or links to executables) a not full install trees, like the SDK.
But i'm open to add any location user's find useful, as the Zephyr SDK cmake packages is about making live easier, then I would like to hear more thoughts on this.

carlescufi replied:
Well, I am in favor of adding it. It doesn't harm (perhaps a few ms wasted on looking for the package there) and it will benefit many? users.

tejlmand replied:
I'm not so fond of it, but i'm overruled by my own fondness of keeping our users happy.
so it's in.

carlescufi commented on doc/getting_started/index.rst:
```diff
@@ -346,23 +346,32 @@ including: compiler, assembler, linker, and their dependencies.
+            * ``/usr/zephyr-sdk[-x.y.z]``
+            * ``/usr/local/zephyr-sdk[-x.y.z]``
+
+	    where ``[-x.y.z]`` is optional text, and can be any text, for example ``-0.11.2``.
```
Add a note that links to the extended info when you choose to install in a different location

tejlmand replied:
yep, good point.

tejlmand replied:
done.

carlescufi reviewed (approved):
This looks good to me as-is, but I've added a few comments for consideration.

carlescufi commented:
@tejlmand 
We discussed this in dev-review (ad-hoc, was not part of the agenda).

This is currently not working with 0.11.1 or 0.11.2. I assume this is intended, but you set the required SDK as 0.11.1, so the error message is quite confusing:

```
CMake Error at /home/carles/src/zephyr/zephyr/cmake/toolchain/zephyr/host-tools.cmake:75 (message):
  The SDK version you are using is not supported, please update your SDK.

  You need SDK version 0.11.1 or newer.

  You have version(s):



  The SDK can be downloaded from:


  https://github.com/zephyrproject-rtos/sdk-ng/releases/download/v0.11.1/zephyr-sdk-0.11.1-setup.run
```

Is there a way to enable this change while at the same time maintaining compatibility for 0.11.x? It would help greatly in introducing the change. If not, then we need to plan for a forced SDK update.

tejlmand commented:
> This is currently not working with 0.11.1 or 0.11.2. I assume this is intended, but you set the required SDK as 0.11.1, so the error message is quite confusing:
> 

As long as I don't know the exact version number when this change will be introduced, then I just had to pick a number.
If the the SDK where this feature is introduced is numbered `0.12`, then that would of course could be the minimum required version.

> Is there a way to enable this change while at the same time maintaining compatibility for 0.11.x? It would help greatly in introducing the change. If not, then we need to plan for a forced SDK update.

That should be possible.
But compatibility with `0.11.1` and `0.11.2` will require us to keep the files that were moved from Zephyr into the SDK to be located in both places.
In Zephyr repo for compatibility with `0.11.1` and `0.11.2`, but then also in the updated `0.11.3` where the CMake package is introduced.
But as soon as we bump the required version to `0.12.x` then we can remove those files from Zephyr repo, and solely have them with the Zephyr SDK.
I will make the required changes and test them out to ensure full compatibility.

Of course user will need the `0.11.3` SDK if he wants to make use of this feature, but it will not force users to update.

So I think this is a good improvement.

tejlmand committed:
cmake: Zephyr sdk backward compatibility with 0.11.1 and 0.11.2

This commit introduces backward compatibility with Zephyr SDK 0.11.1
and 0.11.2 so that users having one of those versions installed can
continue to use that version.

This remove the need to force users to update their SDK.

This is kept in independent commit to ensure it can easily be reverted
when minimum required Zephyr SDK is bumped to version 0.12.

Signed-off-by: Torsten Rasmussen <Torsten.Rasmussen@nordicsemi.no>

tejlmand commented:
> Is there a way to enable this change while at the same time maintaining compatibility for 0.11.x? It would help greatly in introducing the change. If not, then we need to plan for a forced SDK update.

Backward compatibility is now ensured in this commit:
https://github.com/zephyrproject-rtos/zephyr/pull/24143/commits/7ae4f6426681b366cd18d58eefe883fba326d21c
which can easily be reverted when minimum required SDK version becomes 0.12.




> >  This is currently not working with 0.11.1 or 0.11.2. I assume this is intended, but you set the required SDK as 0.11.1, so the error message is quite confusing:
>
> As long as I don't know the exact version number when this change will be introduced, then I just had to pick a number.

I now picked `0.11.3` when checking for Zephyr SDK as CMake package, but code is also validating against `0.11.1` and `0.11.2` if `ZEPHYR_SDK_INSTALL_DIR` and `ZEPHYR_TOOLCHAIN_VARIANT` is set to ensure backward compatibility.

Also changed the error message a bit, so that it will state minimum required is `0.11.1` (which is correct), but the download link will point to `0.11.3` so user gets the latest version.

carlescufi commented:
@tejlmand 
```
  Assertion failed: The detected dtc version is unsupported.

      The version was found to be 1.4.5                                   
      But the minimum supported version is 1.4.6  
```

tejlmand commented:
> @tejlmand
> 
> ```
>   Assertion failed: The detected dtc version is unsupported.
> 
>       The version was found to be 1.4.5                                   
>       But the minimum supported version is 1.4.6  
> ```

Fixed.

mbolivar-nordic reviewed (commented):
Just a couple of minor docs things

mbolivar-nordic commented:
>   Assertion failed: The detected dtc version is unsupported.

This is wrong in general since dtc is optional. I've posted https://github.com/zephyrproject-rtos/zephyr/pull/24824 taking care of that.

tejlmand commented:
> 
> 
> > Assertion failed: The detected dtc version is unsupported.
> 
> This is wrong in general since dtc is optional. I've posted #24824 taking care of that.

but in this case it helped detect that I only reverted 99% of the old behavior, but missed to revert `TOOLCHAIN_ARCH` in the process.
Thanks for this fix.

galak commented:
What's the rule if there are multiple matches in a location?

For example:
```
[galak@spiff build]$ ls -lsd ${HOME}/zephyr-sdk*
     4 drwxr-xr-x. 13 galak galak      4096 May  1 09:20 /home/galak/zephyr-sdk
     4 drwxr-xr-x. 12 galak galak      4096 Apr 22 12:02 /home/galak/zephyr-sdk-0.11.2-pr-204
     4 drwxr-xr-x. 12 galak galak      4096 Apr 23 18:45 /home/galak/zephyr-sdk-0.11.2-pr-206
```

Which one would get picked?

tejlmand commented:
> What's the rule if there are multiple matches in a location?
> 
> For example:
> 
> ```
> [galak@spiff build]$ ls -lsd ${HOME}/zephyr-sdk*
>      4 drwxr-xr-x. 13 galak galak      4096 May  1 09:20 /home/galak/zephyr-sdk
>      4 drwxr-xr-x. 12 galak galak      4096 Apr 22 12:02 /home/galak/zephyr-sdk-0.11.2-pr-204
>      4 drwxr-xr-x. 12 galak galak      4096 Apr 23 18:45 /home/galak/zephyr-sdk-0.11.2-pr-206
> ```
> 
> Which one would get picked?

Under the assumption that all three above matches the minimum requirement, that is `0.11.3` (assume 0.11.2 is a typo) then the first that will be tested is:
`/home/galak/zephyr-sdk`
The order is described here: 
https://cmake.org/cmake/help/v3.17/variable/CMAKE_FIND_PACKAGE_SORT_ORDER.html
https://cmake.org/cmake/help/v3.17/variable/CMAKE_FIND_PACKAGE_SORT_DIRECTION.html
Note that the natural order uses the principle described by `man 3 strverscmp`

If the first package tested does not meet the requirment, then testing continues with the next package, in this case `zephyr-sdk-0.11.2-pr-204`, and then `zephyr-sdk-0.11.2-pr-206`.

See also https://github.com/zephyrproject-rtos/zephyr/blob/759aa1a41461b83f241e2ac057c0d8c2936492cf/cmake/toolchain/zephyr/host-tools.cmake#L35-L44

User can still point to a specific SDK version using `ZEPHYR_SDK_INSTALL_DIR` and `ZEPHYR_TOOLCHAIN_VARIANT.`

galak commented:
> Under the assumption that all three above matches the minimum requirement, that is `0.11.3` (assume 0.11.2 is a typo) then the first that will be tested is:

How is the min ver being checked?  I was playing around with this and it seemed to be ok with 0.11.2 being the version.

tejlmand commented:
> > Under the assumption that all three above matches the minimum requirement, that is `0.11.3` (assume 0.11.2 is a typo) then the first that will be tested is:
> 
> How is the min ver being checked? I was playing around with this and it seemed to be ok with 0.11.2 being the version.

Then you're using the `ZEPHYR_SDK_INSTALL_DIR` and `ZEPHYR_TOOLCHAIN_VARIANT`.

A little more detailed explanation.
We are using `set(MINIMUM_REQUIRED_SDK_VERSION 0.11.3)` and `find_package(Zephyr-sdk ${MINIMUM_REQUIRED_SDK_VERSION} ... )`
https://github.com/zephyrproject-rtos/zephyr/blob/759aa1a41461b83f241e2ac057c0d8c2936492cf/cmake/toolchain/zephyr/host-tools.cmake#L11
https://github.com/zephyrproject-rtos/zephyr/blob/759aa1a41461b83f241e2ac057c0d8c2936492cf/cmake/toolchain/zephyr/host-tools.cmake#L47-L49
in order to find the Zephyr SDK as a CMake package.

If a Zephyr SDK is not found as a CMake package (user could be using a `0.11.1` or `0.11.2`), then we fallback to the backward compatibility mode, and tests the Zephyr SDK version found in `ZEPHYR_SDK_INSTALL_DIR`, to see if that matches: `set(REQUIRED_SDK_VER 0.11.1)`
as seen:
https://github.com/zephyrproject-rtos/zephyr/blob/759aa1a41461b83f241e2ac057c0d8c2936492cf/cmake/toolchain/zephyr/host-tools.cmake#L13-L14
https://github.com/zephyrproject-rtos/zephyr/blob/759aa1a41461b83f241e2ac057c0d8c2936492cf/cmake/toolchain/zephyr/host-tools.cmake#L70-L100


That way we ensure that systems using `0.11.1` or `0.11.2` are still working, while at the same time prefer the CMake package for systems where future SDKs are available.

tejlmand commented:
rebased as:  #23977 is now merged.

tejlmand commented:
have been testing the PR with the 0.11.3-beta-1.
Working as expected, so removing `DNM`

carlescufi commented:
@mbolivar-nordic, @galak or @nashif could we get a second approval on this one?

pfalcon commented:
FYI, apparently this change broken integration with 3rd-party projects (Zephyr JerryScipt and MicroPython ports):

* https://ci.linaro.org/view/lite-iot-ci/job/lite-aeolus-micropython/1181/PLATFORM=qemu_x86,ZEPHYR_GCC_VARIANT=zephyr,label=docker-xenial-amd64-13/console 
* https://ci.linaro.org/view/lite-iot-ci/job/lite-aeolus-jerryscript/PLATFORM=qemu_x86,ZEPHYR_GCC_VARIANT=zephyr,label=docker-xenial-amd64-13/1076/console

Funny thing is that I can't reproduce locally. Will investigate next week.

pfalcon commented:
> FYI, apparently this change broken integration with 3rd-party projects (Zephyr JerryScipt and MicroPython ports):

As a follow up, turned out our builds still used legacy ZEPHYR_GCC_VARIANT variable (as even can be seen links above). Seems like this or nearby patch broke compatibility with that var.

My interest for supporting ZEPHYR_GCC_VARIANT as an alias for newer ZEPHYR_TOOLCHAIN_VARIANT was bisecting, it was quite boring to switch back and forth between 2 vars while doing bisects across Zephry versions. Well, ZEPHYR_TOOLCHAIN_VARIANT was introduced quite some time ago, and hopefully we won't need to bisect that far away. So, RIP ZEPHYR_GCC_VARIANT.

tejlmand commented:
> My interest for supporting ZEPHYR_GCC_VARIANT as an alias for newer ZEPHYR_TOOLCHAIN_VARIANT was bisecting, it was quite boring to switch back and forth between 2 vars while doing bisects across Zephry versions. Well, ZEPHYR_TOOLCHAIN_VARIANT was introduced quite some time ago, and hopefully we won't need to bisect that far away. So, RIP ZEPHYR_GCC_VARIANT.

then look forward to the future, cause with the new CMake package, you can have multiple Zephyr SDK toolchains installed, and not have to change `ZEPHYR_SDK_INSTALL_DIR` when jumping between required versions.
Although it it will take some time before you realize, cause old Zephyr versions still relies old way.
We can't change the past.
See also: https://github.com/zephyrproject-rtos/sdk-ng/pull/199#issue-400121264 description.

pfalcon commented:
> cause with the new CMake package, you can have multiple Zephyr SDK toolchains installed, and not have to change ZEPHYR_SDK_INSTALL_DIR when jumping between required versions.

Thanks, but I have no problem with changing ZEPHYR_SDK_INSTALL_DIR. Actually, it's the most obvious and most trivial way to do it, and I would do it going forward regardless of any other features available to deal with it. My concern and pledge is thus: "please don't introduce regressions [to how people already use it] with this work" ;-).

> Although it it will take some time before you realize, cause old Zephyr versions still relies old way.

Yep, that's what I meant with the bisecting usecase.

Thanks for the reply!

tejlmand commented:
> 
> 
> > cause with the new CMake package, you can have multiple Zephyr SDK toolchains installed, and not have to change ZEPHYR_SDK_INSTALL_DIR when jumping between required versions.
> 
> Thanks, but I have no problem with changing ZEPHYR_SDK_INSTALL_DIR. Actually, it's the most obvious and most trivial way to do it, and I would do it going forward regardless of any other features available to deal with it. My concern and pledge is thus: "please don't introduce regressions [to how people already use it] with this work" ;-).
> 

Actually, i'm a bit unsure why the `ZEPHYR_GCC_VARIANT` breaks your build, cause this change is supposed to be fully backward compatible.
You can still use both `ZEPHYR_SDK_INSTALL_DIR` and `ZEPHYR_TOOLCHAIN_VARIANT=zephyr`, and it will overrule the package mechanism.
Only if `ZEPHYR_TOOLCHAIN_VARIANT` is unset, then it will default to `zephyr`.
Having both `ZEPHYR_TOOLCHAIN_VARIANT` and `ZEPHYR_GCC_VARIANT` should be ok.

So this change should only improve, and not remove any currently supported ways.

pfalcon commented:
> Actually, i'm a bit unsure why the ZEPHYR_GCC_VARIANT breaks your build, cause this change is supposed to be fully backward compatible.

Me neither. And I usually would be concerned and proceed to report it as a regression (by submitting a standalone ticket) and investigate, but given that the only real usecase behind still supporting ZEPHYR_GCC_VARIANT was bisecting, I guess we can relax on that.

> So this change should only improve, and not remove any currently supported ways.

Great, hopefully it will be the same for other cmake-related changes too!
